### DISCUSSION TITLE
How could I use git town with multiple project branches

### DISCUSSION THREAD

**Discussion by dohomi** [git-town/git-town#3988] [General]
Hi Guys,

I am very interested to try Git Town in one of my projects. I would like to know what would be the best strategy for following setup:

develop branch
- every dev starts to hack into this branch and sync and ships

master branch
- as soon I want to make a new release on my production server I push the code into master

project branches
- I have multiple projects set up with a variation of different env variables but they all have the same codebase. I also want to merge into these from develop branch after every release

I would like to make sure, that the master and several project branches are only set up in the way that they can be merged into but never been worked on.

What would be the right way of doing this? I considered to use multiple remotes, each for master and other project branches and only push the develop into each remote.

Thanks for any insight.

**Comment by kevgo** [comment]
Client-side tools like Git Town can help your developers follow best practices, but they can’t enforce rules. Enforcement needs to happen server-side.

If you’re using GitHub, make your `master` branch a [protected branch](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/about-protected-branches). Enable "Require a pull request before merging" and "Require status checks to pass before merging" (and any other checks you find useful) to keep your master branch green. Consider enabling "Require review from Code Owners" and setting up [CODEOWNER files](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) to ensure only authorized people can approve merges. The “Do not allow bypassing the above settings” option is useful to enforce these rules even for admins.

For project branches, it’s easiest to have one Git repo for all projects, with one perennial branch per project. Make these branches protected branches similar to your `master` branch to prevent developers from pushing commits that didn't go through a pull request. Set up GitHub Actions that get triggered when one of your perennial branches receives new commits and deploy to the respective cloud environment. Then you can merge new commits from the `master` branch into the perennial branches via pull requests.

Using pull requests for deployments has several benefits:

- you can tag people who should be aware of the deploy
- the pull request for the deploy is a good place to discuss this specific deploy
- the GitHub action can comment on the pull request when the deploy is finished

This setup should be secure enough. If you need even stronger protection, or different user accounts per project, then you need to set up multiple projects and link them via remotes.

On the client side, Git Town can make it easy for your developers to do the right things. If you commit a [Git Town configuration file](https://www.git-town.com/configuration-file) into your repo, these settings automatically apply to all developers who use Git Town.

- make `master` the [main branch](https://www.git-town.com/branch-types#main-branch)
- make the project branches [perennial branches](https://www.git-town.com/branch-types#perennial-branches)
- use [git town propose](https://www.git-town.com/commands/propose) to create pull requests
- consider using GitHub's [merge queue](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/merging-a-pull-request-with-a-merge-queue) to automatically merge pull request once tests pass

If you go the multiple-remote way, Git Town's ability to [sync from the upstream remote](https://www.git-town.com/commands/sync.html?highlight=upstream#configuration) could be useful.

Does this make sense?

**Comment by dohomi** [comment]
Hi @kevgo that is a great explanation, I was already thinking that perennial branches might be the right way of doing it with Git Town.

I will start setting up my project with git town and see how far I can go. As I use Nhost (backend) and Vercel (frontend) both support direct deploy via Git branches so your proposed ideas make total sense for my setup.

**Reply by kevgo** [reply]
Good luck, and reach out if you have any question!
